Add billing related queries and documentation

oapa opened the pull request.

ryanmcdowell commented on scripts/billing/sql/daily_compute_cores_average.sql:
```diff
@@ -0,0 +1,17 @@
+  `data-analytics-pocs.public.billing_dashboard_export`,
+  UNNEST(system_labels) AS system_labels
+WHERE 
+  service.id = "6F81-5844-456A"
```
Single-quotes

ryanmcdowell replied:
Can you add 2 spaces after the quotes and a single-line comment describing what this identifier is?

Example:
```sql
WHERE 
  service.id = '6F81-5844-456A'  -- Compute Engine
```

ryanmcdowell commented on scripts/billing/sql/daily_compute_cores_average.sql:
```diff
@@ -0,0 +1,17 @@
+  UNNEST(system_labels) AS system_labels
+WHERE 
+  service.id = "6F81-5844-456A"
+  AND system_labels.key = "compute.googleapis.com/cores"
```
Single-quotes

ryanmcdowell commented on scripts/billing/sql/daily_compute_discount.sql:
```diff
@@ -0,0 +1,15 @@
+*/
+
+SELECT
+  FORMAT_DATETIME("%Y%m%d", DATETIME(usage_end_time)) AS date_ymd, 
```
Single-quotes

ryanmcdowell commented on scripts/billing/sql/daily_compute_discount.sql:
```diff
@@ -0,0 +1,15 @@
+FROM
+  `data-analytics-pocs.public.billing_dashboard_export`,
+  UNNEST(credits) AS credits
+WHERE service.id = "6F81-5844-456A"
```
Can you add 2 spaces after the quotes and a single-line comment describing what this identifier is?

Example:
```sql
WHERE 
  service.id = '6F81-5844-456A'  -- Compute Engine
```

ryanmcdowell commented on scripts/billing/sql/daily_compute_usage_hours.sql:
```diff
@@ -0,0 +1,23 @@
+*/
+
+SELECT
+  FORMAT_DATETIME("%Y%m%d", DATETIME(usage_end_time)) AS date_ymd, 
```
Single-quotes

ryanmcdowell commented on scripts/billing/sql/daily_compute_usage_hours.sql:
```diff
@@ -0,0 +1,23 @@
+FROM
+  `data-analytics-pocs.public.billing_dashboard_export`,
+  UNNEST(credits) AS credits
+WHERE service.id = "6F81-5844-456A"
```
Can you add 2 spaces after the quotes and a single-line comment describing what this identifier is?

Example:
```sql
WHERE 
  service.id = '6F81-5844-456A'  -- Compute Engine
```

ryanmcdowell commented on scripts/billing/sql/forecast_net_balance.sql:
```diff
@@ -0,0 +1,114 @@
+  month,
+  balance,
+  future_balance
+FROM (
```
Sub-selects parens should be placed on a new line and indented by 2-spaces ([http://go/sqlguide#clause-block](http://go/sqlguide#clause-block))

Please also apply to all the sub-selects below as well. If heavily nested, think about using the `WITH` clause for greater readability.

ryanmcdowell commented on scripts/billing/sql/forecast_net_balance.sql:
```diff
@@ -0,0 +1,114 @@
+      + EXTRACT(MONTH FROM CURRENT_TIMESTAMP())) + INTERCEPT AS future_balance
+  FROM (
+    SELECT 
+      SLOPE,
```
In this `SELECT` clause and ones below, the columns are aliased with UPPER_CASE but other files and within this file, columns are aliased with lower_case. Please change the UPPER_CASE aliases to lower_case to be consistent.

ryanmcdowell commented on scripts/billing/sql/forecast_net_balance.sql:
```diff
@@ -0,0 +1,114 @@
+          GROUP BY X
+          ORDER BY X
+        )
+        WHERE X IS NOT NULL AND
```
Place the first condition on a new-line and the `AND` clause on the same line as the second condition.

ryanmcdowell commented on scripts/billing/sql/monthly_costs_credits.sql:
```diff
@@ -0,0 +1,12 @@
+  ROUND(SUM((SELECT SUM(amount) FROM UNNEST(credits))),2) AS credits
+FROM `data-analytics-pocs.public.billing_dashboard_export`
+GROUP BY year, month
+ORDER by year, month
```
Missing end of file new-line

ryanmcdowell commented on scripts/billing/sql/monthly_invoice_costs_credits.sql:
```diff
@@ -0,0 +1,15 @@
+/*
+Query returns the invoice total for each month, as a sum of regular costs,
```
Remove comma after month

ryanmcdowell commented on scripts/billing/sql/monthly_invoice_costs_credits.sql:
```diff
@@ -0,0 +1,15 @@
+SELECT
+  invoice.month,
+  SUM(cost) + SUM(IFNULL((SELECT SUM(c.amount) FROM UNNEST(credits) c), 0)) AS total,
+  (SUM(CAST(cost * 1000000 AS int64))
```
Capitalize keyword `INT64` here and below

ryanmcdowell commented on scripts/billing/sql/monthly_invoice_costs_credits.sql:
```diff
@@ -0,0 +1,15 @@
+  invoice.month,
+  SUM(cost) + SUM(IFNULL((SELECT SUM(c.amount) FROM UNNEST(credits) c), 0)) AS total,
+  (SUM(CAST(cost * 1000000 AS int64))
+    + SUM(IFNULL((SELECT SUM(CAST(c.amount * 1000000 as int64))
```
This would be cleaner if the `SELECT` was on one line.

ryanmcdowell commented on scripts/billing/sql/monthly_invoice_costs_credits.sql:
```diff
@@ -0,0 +1,15 @@
+                  FROM UNNEST(credits) c), 0))) / 1000000
+    AS total_exact
+FROM `data-analytics-pocs.public.billing_dashboard_export`
+GROUP BY 1
```
Prefer explicit column names

ryanmcdowell commented on scripts/billing/sql/previous_day_costs_by_project.sql:
```diff
@@ -0,0 +1,11 @@
+  sum(cost) AS costs 
+FROM `data-analytics-pocs.public.billing_dashboard_export` 
+WHERE DATE(export_time) = DATE_SUB(CURRENT_DATE(), INTERVAL 1 DAY)
+GROUP BY 1
```
Prefer explicit column names

ryanmcdowell commented on scripts/billing/sql/previous_day_costs_by_project.sql:
```diff
@@ -0,0 +1,11 @@
+*/
+
+SELECT
+  project.name,
```
Should this be aliased?

```sql
AS project_name
```

ryanmcdowell commented on scripts/billing/sql/previous_day_costs_by_service.sql:
```diff
@@ -0,0 +1,12 @@
+  sum(cost) AS costs
+FROM `data-analytics-pocs.public.billing_dashboard_export` 
+WHERE DATE(export_time) = DATE_SUB(CURRENT_DATE(), INTERVAL 1 DAY)
+GROUP BY 1, 2
```
Prefer explicit column names

ryanmcdowell commented on scripts/billing/sql/previous_month_costs_by_label.sql:
```diff
@@ -0,0 +1,19 @@
+LEFT JOIN UNNEST(labels) as labels
+  ON labels.key = "service"
+WHERE invoice.month = FORMAT_DATE("%Y%m", DATE_SUB(CURRENT_DATE(), INTERVAL 1 MONTH))
+# for all invoice months, uncomment below (and comment above)
```
Use `--` for single-line comments for consistency with the rest of the repo.

ryanmcdowell commented on scripts/billing/sql/previous_month_costs_by_label.sql:
```diff
@@ -0,0 +1,19 @@
+FROM `data-analytics-pocs.public.billing_dashboard_export`
+LEFT JOIN UNNEST(labels) as labels
+  ON labels.key = "service"
+WHERE invoice.month = FORMAT_DATE("%Y%m", DATE_SUB(CURRENT_DATE(), INTERVAL 1 MONTH))
```
Single-quotes

ryanmcdowell commented on scripts/billing/sql/previous_day_costs_by_service.sql:
```diff
@@ -0,0 +1,12 @@
+*/
+
+SELECT
+  service.id,
```
We should probably alias the id and description fields as done in other files within this PR.

```sql
  service.id AS service_id,
  service.description AS service_description
```

ryanmcdowell commented on scripts/billing/sql/previous_month_costs_by_label.sql:
```diff
@@ -0,0 +1,19 @@
+  invoice.month as invoice_month,
+  labels.value as service,
+  SUM(cost) as costs,
+  ROUND(SUM((SELECT SUM(amount) FROM UNNEST(credits))),2) as credits 
```
Add space between `,` and `2`

ryanmcdowell commented on scripts/billing/sql/previous_month_costs_by_service.sql:
```diff
@@ -0,0 +1,14 @@
+  service.id AS service_id,
+  service.description AS service_description,
+  ROUND(SUM(cost), 2) AS costs,
+  ROUND(SUM((SELECT SUM(amount) FROM UNNEST(credits))),2) AS credits 
```
Add space between `,` and `2`

ryanmcdowell commented on scripts/billing/sql/previous_month_costs_by_service.sql:
```diff
@@ -0,0 +1,14 @@
+  ROUND(SUM(cost), 2) AS costs,
+  ROUND(SUM((SELECT SUM(amount) FROM UNNEST(credits))),2) AS credits 
+FROM `data-analytics-pocs.public.billing_dashboard_export`
+WHERE invoice.month = FORMAT_DATE("%Y%m", DATE_SUB(CURRENT_DATE(), INTERVAL 1 MONTH))
```
Single-quotes

ryanmcdowell commented on scripts/billing/sql/previous_month_costs_by_service_label.sql:
```diff
@@ -0,0 +1,21 @@
+  labels.value AS label_value,
+  service.description AS description,
+  SUM(cost) AS costs,
+  ROUND(SUM((SELECT SUM(amount) FROM UNNEST(credits))),2) AS credits 
```
Add space between `,` and `2`

ryanmcdowell commented on scripts/billing/sql/previous_month_costs_by_service_label.sql:
```diff
@@ -0,0 +1,21 @@
+  ROUND(SUM((SELECT SUM(amount) FROM UNNEST(credits))),2) AS credits 
+FROM `data-analytics-pocs.public.billing_dashboard_export`
+LEFT JOIN UNNEST(labels) AS labels
+  ON labels.key = "cost_center"
```
Single-quotes here and below

ryanmcdowell commented on scripts/billing/sql/total_costs_by_sku_label.sql:
```diff
@@ -0,0 +1,10 @@
+  TO_JSON_STRING(labels) AS labels,
+  sum(cost) AS cost
+FROM `data-analytics-pocs.public.billing_dashboard_export`
+GROUP BY 1,2
```
Prefer explicit column names

oapa replied:
Updated group by and order to use explicit column names here and across files

ryanmcdowell commented on scripts/billing/sql/total_costs_by_sku_label.sql:
```diff
@@ -0,0 +1,10 @@
+*/
+
+SELECT
+  sku.description,
```
Alias this field

```sql
  sku.description as sku_description
```

oapa replied:
Aliased this field and similar ones across files

ryanmcdowell commented on scripts/billing/README.md:
```diff
@@ -0,0 +1,31 @@
+
+In addition to the Billing product available within the GCP Console, you may want to export billing data to BigQuery for custom analytics.  Below are some example queries to help get you started using the standard data export format.  See the [links below](#useful-links) to learn more about GCP billing datasets including how to export your own billing data for analysis.
+
+The example SQL snippets below reference a sample billing export available as a public dataset in BigQuery (**`data-analytics-pocs.public.billing_dashboard_export`**).  You will need to replace references to this table with your own dataset. 
```
Change to `bqutil.billing` once sample data is migrated.

oapa replied:
Migrated dataset and updated queries/readme

ryanmcdowell commented on scripts/billing/README.md:
```diff
@@ -0,0 +1,31 @@
+
+In addition to the Billing product available within the GCP Console, you may want to export billing data to BigQuery for custom analytics.  Below are some example queries to help get you started using the standard data export format.  See the [links below](#useful-links) to learn more about GCP billing datasets including how to export your own billing data for analysis.
+
+The example SQL snippets below reference a sample billing export available as a public dataset in BigQuery (**`data-analytics-pocs.public.billing_dashboard_export`**).  You will need to replace references to this table with your own dataset. 
```
"... your own **project and** dataset."

freedomofnet commented on scripts/billing/sql/previous_month_costs_by_label.sql:
```diff
@@ -0,0 +1,19 @@
+This query uses a label key of "service" as an example
+*/
+
+SELECT
```
SELECT
  invoice.month AS invoice_month,
  labels.value AS service,
  labels.key AS key,
  SUM(cost) AS costs,
  ROUND(SUM(credits.amount), 2) as credits 
FROM
  `data-analytics-pocs.public.billing_dashboard_export`
  LEFT JOIN UNNEST(credits) AS credits
  LEFT JOIN UNNEST(labels) as labels
WHERE invoice.month = FORMAT_DATE("%Y%m", DATE_SUB(CURRENT_DATE(), INTERVAL 1 MONTH))
AND labels.key = 'service'
-- for all invoice months, uncomment below (and comment above)
-- WHERE invoice.month IS NOT null
GROUP BY
  invoice.month,
  labels.value,
  labels.key;

oapa replied:
Updated to use the provided pattern

oapa replied:
Updated with this pattern

freedomofnet commented on scripts/billing/sql/previous_month_costs_by_service_label.sql:
```diff
@@ -0,0 +1,21 @@
+This example uses a label key of "cost_center" and label value of "sales"
+*/
+
+SELECT
```
Please refer to previous_month_costs_by_label.sql for similar pattern.

oapa replied:
Updated to follow the pattern provided

oapa replied:
Updated to use similar pattern

freedomofnet commented on scripts/billing/sql/total_costs_by_label_value.sql:
```diff
@@ -0,0 +1,13 @@
+This example uses a label key of "environment" as an example
+*/
+
+SELECT
```
This query will return costs for labels.key is NULL, is it intended.

If not, please put the filter labels.key to where clause.

oapa replied:
Nice catch. Updated to filter NULL keys

oapa replied:
Added filter for null key values

oapa commented:
Thank you for the feedback! I've updated the queries based on the comments and style guide.  I've also moved the sample dataset into the "bqutil" project and updated docs/queries accordingly.

ryanmcdowell commented:
> Thank you for the feedback! I've updated the queries based on the comments and style guide. I've also moved the sample dataset into the "bqutil" project and updated docs/queries accordingly.

Can you please resolve the individual comments? I think you may have missed some in your latest update.

oapa commented on scripts/billing/sql/total_costs_by_label_value.sql:
```diff
@@ -0,0 +1,13 @@
+  SUM(cost) AS costs
+FROM `data-analytics-pocs.public.billing_dashboard_export`
+LEFT JOIN UNNEST(labels) AS labels
+  ON labels.key = "environment"
```
Updated all double quotes to single quotes across files

oapa commented on scripts/billing/sql/daily_compute_cores_average.sql:
```diff
@@ -0,0 +1,17 @@
+*/
+
+SELECT
+  FORMAT_DATETIME("%Y%m%d", DATETIME(usage_end_time)) AS date_ymdh, 
```
Updated

oapa commented on scripts/billing/sql/daily_compute_cores_average.sql:
```diff
@@ -0,0 +1,17 @@
+
+SELECT
+  FORMAT_DATETIME("%Y%m%d", DATETIME(usage_end_time)) AS date_ymdh, 
+  AVG(CAST(system_labels.value as int64)) AS avg_core_usage
```
Keywords capitalized across all files

oapa commented on scripts/billing/sql/daily_compute_usage_hours.sql:
```diff
@@ -0,0 +1,23 @@
+  sku.description as component,
+  SUM(usage.amount) as component_usage,
+  usage.unit as usage_unit,
+  ROUND(SUM(cost - credits.amount) * 100) / 100 AS actual_cost
```
Updated here and across files

oapa commented on scripts/billing/sql/forecast_net_balance.sql:
```diff
@@ -0,0 +1,114 @@
+  FROM `data-analytics-pocs.public.billing_dashboard_export`
+  WHERE
+    FORMAT_TIMESTAMP('%Y%m', usage_end_time) 
+      >= FORMAT_DATE('%Y%m', DATE_ADD(CURRENT_DATE(), INTERVAL -3 MONTH))
```
Makes sense — updated with that pattern

oapa commented on scripts/billing/sql/forecast_net_balance.sql:
```diff
@@ -0,0 +1,114 @@
+        CORRELATION
+      FROM (
+        SELECT 
+          COUNT(*) AS N,
```
While modifying, I had some questions about the logic in this query. Rolling back for now while I review..

oapa commented:
I resolved the individual comments and removed the forecasting query from this commit.  Upon further review, I think there are logic issues that need to be fixed.
